src/rnp/rnpcfg.c

```diff
@@ -0,0 +1,502 @@
+void
+rnp_cfg_free(rnp_cfg_t *cfg)
+{
+    int i;
+
+    for (i = 0; i < cfg->count; i++) {
+        free(cfg->vals[i]);
+        free(cfg->keys[i]);
+    }
+
+    free(cfg->keys);
+    free(cfg->vals);
+}
+
+int
+rnp_cfg_get_pswdtries(rnp_cfg_t *cfg)
+{
+    const char *numtries;
+    int         num;
+
+    numtries = rnp_cfg_get(cfg, CFG_NUMTRIES);
+
+    if ((numtries == NULL) || ((num = atoi(numtries)) <= 0)) {
+        return MAX_PASSPHRASE_ATTEMPTS;
+    } else if (strcmp(numtries, "unlimited")) {
+        return INFINITE_ATTEMPTS;
+    } else {
+        return num;
+    }
+}
+
+int
+rnp_cfg_check_homedir(rnp_cfg_t *cfg, char *homedir)
+{
+    struct stat st;
+    int         ret;
+
+    if (homedir == NULL) {
+        fputs("rnp: homedir option and HOME environment variable are not set \n", stderr);
+        return RNP_FAIL;
+    } else if ((ret = stat(homedir, &st)) == 0 && !S_ISDIR(st.st_mode)) {
+        /* file exists in place of homedir */
+        fprintf(stderr, "rnp: homedir \"%s\" is not a dir\n", homedir);
+        return RNP_FAIL;
+    } else if (ret != 0 && errno == ENOENT) {
+        /* If the path doesn't exist then fail. */
+        fprintf(stderr, "rnp: warning homedir \"%s\" not found\n", homedir);
+        return RNP_FAIL;
+    } else if (ret != 0) {
+        /* If any other occurred then fail. */
+        fprintf(stderr, "rnp: an unspecified error occurred\n");
+        return RNP_FAIL;
+    }
+
+    return RNP_OK;
+}
+
+/* read any gpg config file */
+static int
+conffile(const char *homedir, char *userid, size_t length)
+{
+    regmatch_t matchv[10];
+    regex_t    keyre;
+    char       buf[BUFSIZ];
+    FILE *     fp;
+
+    (void) snprintf(buf, sizeof(buf), "%s/.gnupg/gpg.conf", homedir);
+    if ((fp = fopen(buf, "r")) == NULL) {
+        return RNP_FAIL;
+    }
+    (void) memset(&keyre, 0x0, sizeof(keyre));
+    (void) regcomp(&keyre, "^[ \t]*default-key[ \t]+([0-9a-zA-F]+)", REG_EXTENDED);
+    while (fgets(buf, (int) sizeof(buf), fp) != NULL) {
+        if (regexec(&keyre, buf, 10, matchv, 0) == 0) {
+            (void) memcpy(userid,
+                          &buf[(int) matchv[1].rm_so],
+                          MIN((unsigned) (matchv[1].rm_eo - matchv[1].rm_so), length));
+
+            (void) fprintf(stderr,
+                           "rnp: default key set to \"%.*s\"\n",
+                           (int) (matchv[1].rm_eo - matchv[1].rm_so),
+                           &buf[(int) matchv[1].rm_so]);
+        }
+    }
+    (void) fclose(fp);
+    regfree(&keyre);
+    return RNP_OK;
+}
+
+/*
+  Compose path from dir, subdir and filename. subdir can be null, then just dir and filename
+  will be used.
+  res should point to the allocated buffer.
+*/
+int
+rnp_path_compose(const char *dir, const char *subdir, const char *filename, char *res)
+{
+    int pos;
+
+    /* checking input parameters for conrrectness */
+    if (!dir || !filename || !res) {
+        return RNP_FAIL;
```

Comment:
@ronaldtse I mean - if this is function that can be called by external library then we should return descriptive error, otherwise we should crash as it's implementation error.